snuba/admin/static/rpc_endpoints/index.tsx

```diff
@@ -1,240 +1,304 @@
-                                    <Text>No query info available</Text>
-                                )}
-                            </Accordion.Panel>
-                        </Accordion.Item>
-                    </Accordion>
-                    <Space h="md" />
-                    <h4>Response Data:</h4>
-                    <Code block>
-                        <pre>
-                            {JSON.stringify(
-                                (({ meta, ...rest }) => rest)(response),
-                                null,
-                                2
-                            )}
-                        </pre>
-                    </Code>
-                </>
-            )}
-        </div>
-    );
+          </Accordion.Panel>
+        </Accordion.Item>
+      </Accordion>
+      <Space h="md" />
+      <Textarea
+        label="Request Body (JSON)"
+        placeholder="Enter request body"
+        value={requestBody}
+        onChange={(event) => setRequestBody(event.currentTarget.value)}
+        style={{ width: '100%' }}
+        autosize
+        minRows={5}
+      />
+      <Space h="md" />
+      <Checkbox
+        label="Enable Debug Mode"
+        checked={debugMode}
+        onChange={(event) => setDebugMode(event.currentTarget.checked)}
+        disabled={!DEBUG_SUPPORTED_VERSIONS.includes(selectedVersion || '')}
+        className={classes.debugCheckbox}
+      />
+      <Button
+        onClick={handleExecute}
+        disabled={!selectedEndpoint || !requestBody || isLoading}
+        leftIcon={isLoading ? <Loader size="xs" /> : null}
+      >
+        {isLoading ? 'Executing...' : 'Execute'}
+      </Button>
+      {response && (
+        <>
+          <Space h="md" />
+          <h3>Response:</h3>
+          <Switch
+            className={classes.viewToggle}
+            label="Show Trace Logs"
+            checked={showTraceLogs}
+            onChange={(event) => setShowTraceLogs(event.currentTarget.checked)}
+          />
+          <h4>Response Metadata:</h4>
+          <Accordion classNames={{ item: classes.accordion }}>
+            <Accordion.Item value="query-info">
+              <Accordion.Control>
+                <Text fw={700}>
+                  {showTraceLogs ? 'Query Trace Logs' : 'Query Metadata'}
+                </Text>
+              </Accordion.Control>
+              <Accordion.Panel>
+                {response.meta?.queryInfo ? (
+                  response.meta.queryInfo.map((queryInfo: QueryInfo, index: number) => (
+                    <div key={index}>
+                      {showTraceLogs ? (
+                        <>
+                          <h4>Trace Logs</h4>
+                          {queryInfo.traceLogs ? (
+                            <div
+                              className={classes.traceLogsContainer}
+                              dangerouslySetInnerHTML={{
+                                __html: format_trace_log(queryInfo.traceLogs)
```

Comment:
no other way to do this?

Reply:
in this specific case, dangerouslySetInnerHTML is being used safely because:
The HTML content is fully controlled by snuba:
The HTML structure is hardcoded (span elements with specific styles)
User input (log messages) is properly escaped using escapeHtml()
Color values are constants defined in your code
The trace logs are presumably coming from snuba, not from user input or external sources

Reply:
I think `format_trace_log` should be changed to return a React component instead of a string. Then this wouldn't be necessary at all, you could do:
```
queryInfo.traceLogs ? (<div>{format_trace_log(queryInfo.traceLogs)}</div>) : ...
```

Reply:
Refactored the code to return a React component instead of a string.